Add optional x_jitter to dependence_plot

ihopethiswillfi opened the pull request:
Optionally add random noise to feature values if feature has less than 20 unique values.  May increase plot readability when feature is non-continuous.

Attached example picture with noise set to 0.33.
Without the noise, most of the dots overlap, especially where X == 2.0
![plot_interaction_businesstravel_vs_age](https://user-images.githubusercontent.com/17611570/48430353-7085c080-e76f-11e8-8ac2-a078b92c117a.png)

ihopethiswillfi committed:
Add optional noise when feature has less than 20 unique values.

Optionally add random noise to feature values if feature has less than 20 unique values.  May increase plot readability when feature is non-continuous.

JacobMcc commented on shap/plots/dependence.py:
```diff
@@ -138,6 +142,15 @@ def convert_name(ind):
+    if (noise > 0) and (len(set(xv)) < 20):
+        if noise > 1: noise = 1
+        xvals = list(set(xv))
+        smallest_diff = np.min([xvals[i] - xvals[i-1] for i in range(1, len(xvals))])
```
```suggestion
        smallest_diff = np.min(np.diff(np.sort(xvals)))
```
You could use `np.min(np.diff(xval))` here to avoid the comprehension. Also `set(xv)` may not preserve ordering, so you may want to call `sort` on the `xvals`. If `xvals` isn't ordered then you might end up with a negative value for `smallest_diff`.

JacobMcc commented on shap/plots/dependence.py:
```diff
@@ -138,6 +142,15 @@ def convert_name(ind):
+        xvals = list(set(xv))
+        smallest_diff = np.min([xvals[i] - xvals[i-1] for i in range(1, len(xvals))])
+        noise_amount = noise * smallest_diff
+        for i in range(len(xv)):
```
```suggestion
        xv += (np.random.ranf(size = len(xv))*jitter_amount) - (jitter_amount/2)
```
You could get ride of this for loop and take advantage of vectorized operations here. 
Overall I like your contribution to the dependence plot btw, great idea!

ihopethiswillfi replied:
Thanks!  Good point on the vectorization.  But I think you should also remove the line just under the FOR statement (cause you replaced it).  Can you do that or let me know how to.  Sorry I'm kinda a git rookie.

JacobMcc replied:
No problem. See comment below.

slundberg commented:
Thanks @ihopethiswillfi ! This looks good, we should name it `x_jitter` though since jitter is the name of this kind of noise and we are only adding it to the x-axis. The vectorization suggestions from @JacobMcc are great, you can also consider using `np.unique`.

One other thing, I am not sure about the limit on 20 unique values. It seems that we shouldn't assume a user doesn't want jitter for 30 values (especially if they make a wider plot). It is also a bit strange to turn on jitter only to have it turn off automatically. If at some point in the future we auto-detect categorical features then we could auto turn on jitter for those.

Once these changes are in I am happy to merge :)

Commit:
remove list comprehension

use np.min(np.diff(xval)) here to avoid the comprehension. Also set(xv) may not preserve ordering, so you may want to call sort on the xvals. If xvals isn't ordered then you might end up with a negative value for smallest_diff.

Co-Authored-By: ihopethiswillfi <ihopethiswillfi@yahoo.com>

ihopethiswillfi commented:
@slundberg Agreed with everything there.  
- will rename to x_jitter
- will remove arbitrary cap of 20.  I also wasn't sure about that either when committing it.  I originally thought I needed this to get what I wanted but I now found a way around it.  I can just detect the number of unique values of a feature before I call dependence_plot and base my jitter decision on that.

ihopethiswillfi committed:
renamed noise to x_jitter

ihopethiswillfi committed:
removed arbitrary jitter cap of 20 unique X values

JacobMcc commented on shap/plots/dependence.py:
```diff
@@ -138,6 +142,15 @@ def convert_name(ind):
+        smallest_diff = np.min(np.diff(np.sort(xvals)))
+        jitter_amount = x_jitter * smallest_diff
+        for i in range(len(xv)):
+            xv[i] += (np.random.ranf(size=1)[0] * jitter_amount) - (jitter_amount/2)
```
```suggestion

```
Here is the deletion of the line previously in the for loop. For future reference you can just change the code on your machine and then commit/push to GH with the changes and the Pull Request will automatically update.

JacobMcc commented on shap/plots/dependence.py:
```diff
@@ -138,6 +142,15 @@ def convert_name(ind):
+    # optionally add jitter to feature values
+    if x_jitter > 0:
+        if x_jitter > 1: x_jitter = 1
+        xvals = list(set(xv))
```
```suggestion
        xvals = np.unique(xv)
```
Per @slundberg 's recommendation

Commit:
Update shap/plots/dependence.py

Here is the deletion of the line previously in the for loop

Co-Authored-By: ihopethiswillfi <ihopethiswillfi@yahoo.com>

Commit:
Update shap/plots/dependence.py

Co-Authored-By: ihopethiswillfi <ihopethiswillfi@yahoo.com>

Commit:
vectorize FOR loop.

vectorize FOR loop.

Co-Authored-By: ihopethiswillfi <ihopethiswillfi@yahoo.com>

slundberg commented:
Thanks! The windows test fail has to do with an unrelated C++ VLA stack issue.

I have been thinking more about this, and since it happens for categorical data perhaps we should just add a flag denoting categorical data. With `categorical="auto"` by default. Then we could jitter automatically for small category counts. This would be similar to your 20 threshold, but it would also then enable us to do interesting things with the x jitter, such as doing non-uniform jitter to show the number of samples in each category (width of the jitter can be proportional to sample counts). This would simultaneously convey histogram information without adding mush visual complexity, since the width of the jitter would be the same as the height of a category histogram. This would be a much more precise and robust way of communicating this information than trying to pick the right opacity so we can tell where the most points are.

Anyway, let me think about this a bit more before merging, because if we go with a `categorical` flag then that would change this api.

slundberg commented:
I am going to merge this. If we add support for categorical features later we can update things then. Thanks again!
